include/boost/geometry/algorithms/detail/buffer/piece_border.hpp

```diff
@@ -0,0 +1,509 @@
+                    && step(point, m_originals[0], m_originals[1], tir, por_original, state)
+                    && step(point, m_originals[1], offsetted_front, tir, por_to_offsetted, state);
+        }
+
+        if (continue_processing)
+        {
+            // Check the offsetted ring (in rounded joins, these might be
+            // several segments)
+            walk_offsetted(point, m_ring->begin() + m_begin, m_ring->begin() + m_end,
+                           tir, state);
+        }
+
+        return true;
+    }
+
+private :
+
+    static geometry::strategy::buffer::place_on_ring_type
+        adapted_place_on_ring(geometry::strategy::buffer::place_on_ring_type target,
+                              bool one_sided, bool is_linear_end_point)
+    {
+        return one_sided || is_linear_end_point
+               ? geometry::strategy::buffer::place_on_ring_offsetted
+               : target;
+    }
+
+    template <typename TurnPoint, typename Iterator, typename Strategy, typename State>
+    bool walk_offsetted(TurnPoint const& point, Iterator begin, Iterator end, Strategy const & strategy, State& state) const
+    {
+        Iterator it = begin;
+        Iterator beyond = end;
+
+        // Move iterators if the offsetted ring is monotonic increasing or decreasing
+        if (m_is_monotonic_increasing)
+        {
+            if (! get_range_around(begin, end, point, geometry::less<Point, 0>(), it, beyond))
+            {
+                return true;
+            }
+        }
+        else if (m_is_monotonic_decreasing)
+        {
+            if (! get_range_around(begin, end, point, geometry::greater<Point, 0>(), it, beyond))
+            {
+                return true;
+            }
+        }
+
+        for (Iterator previous = it++ ; it != beyond ; ++previous, ++it )
+        {
+            if (! step(point, *previous, *it, strategy,
+                 geometry::strategy::buffer::place_on_ring_offsetted, state))
+            {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    template <typename TurnPoint, typename Strategy, typename State>
+    bool step(TurnPoint const& point, Point const& p1, Point const& p2, Strategy const & strategy,
+              geometry::strategy::buffer::place_on_ring_type place_on_ring, State& state) const
+    {
+        // A step between original/offsetted ring is always convex
+        // (unless the join strategy generates points left of it -
+        //  future: convexity might be added to the buffer-join-strategy)
+        // Therefore, if the state count > 0, it means the point is left of it,
+        // and because it is convex, we can stop
+
+        typedef typename geometry::coordinate_type<Point>::type coordinate_type;
+        typedef geometry::detail::distance_measure<coordinate_type> dm_type;
+        dm_type const dm = geometry::detail::get_distance_measure<cartesian_tag>(point, p1, p2);
```

Comment:
Shouldn't this tag be taken from the Strategy?

Reply:
ah, yes, of course. Will fix.